### ISSUE TITLE
Swift workflow loop updates asynchronously, can cause lost events

### ISSUE THREAD

**Issue by davidapgar** [square/workflow#348]
tl;dr: Handling an action and having a render pass occur yields the main loop, which can cause UI events to be missed, as the previous sink will be invalid, and the updated sink won't be set on the UI side.

Longer form example that shows the behavior:
```swift
// The state of the text field is managed in the workflow. Updates to it when edited are sent back via the `.didChange` closure.
struct ExampleScreen: Screen {
  var text: String
  var didChange: (String) -> Void
}

final class ExampleViewController /*...*/ {
  let textField: UITextField

  /* snipped */

    override func viewDidLoad() {
        super.viewDidLoad()

        // Any time the text changes, send the event back to the workflow
        textField.addTarget(
            self,
            action: #selector(textDidChange(sender:)),
            for: .editingChanged)

        // rest of viewDidLoad
    }

    @objc private func textDidChange(sender: UITextField) {
        guard let text = sender.text else { return }
        screen.onTextChanged(text)
    }

    override func screenDidChange(from previousScreen: PlaceholderScreen) {
        // Update the text field state based on the state from the workflow.
        textField.text = screen.text
    }
}
```

Relevant parts of workflow implementation:
```swift
// MARK: Actions

extension RootWorkflow {

    enum Action: WorkflowAction {

        typealias WorkflowType = RootWorkflow

        case updated(String)

        func apply(toState state: inout RootWorkflow.State) -> RootWorkflow.Output? {

            switch self {
            case .updated(let text):
                state.text = text
            }
            return nil
        }

    }
}


// MARK: Rendering

extension RootWorkflow {

    public typealias Rendering = ExampleScreen

    public func render(state: State, context: RenderContext<RootWorkflow>) -> Rendering {
        let sink = context.makeSink(of: Action.self)

        return PlaceholderScreen(
            text: state.text,
            onTextChanged: { text in
                sink.send(.updated(text))
            })
    }
}
```
Explanation of what's happening:
The text field's `textDidChange` action occurs, and the current text value is sent back to the workflow with a `sink.send(.updated(text))` call.

The `apply` method is called, and the workflow's state is updated with the updated text. The next render pass occurs, and the value of `state.text` is apply back to the UITextField.

Because the text input is happening rapidly, UIKit will send the next event to through the text field as soon as the main thread is available. This is easiest to reproduce synthetically with a KIF test, but using a hardware keyboard *can* produce this behavior (just less consistently).

Since the Signal backing the sink is no longer valid after a first action has been received (and it doesn't, nor shouldn't, batch the events), when the second time `sink.send(.updated(text))` is called, the action is dropped.

The text field's state now includes two keypresses (ie: "ab"), but the workflow has only gotten an updated text for the first (ie: "a"). When the render pass completes, the text field (being applied from the screen) will have `textField.text` be "a", and we've lost an entered character.

The most likely path forward for this is to ensure that it is synchronous for the next render pass to occur directly after handling an action (ie: use the UIScheduler in the `SubtreeManager`). This does *not* prevent this in the case where the UI update based on a new screen is happening asynchronously (deferred applying of the updated closure) - so there still could be cases where the UI implementation can cause dropped events.

**Comment by davidapgar** [comment]
(A series of notes and explanations I wrote up while investigating this. Should give more context of what the existing behavior is, and what the proposed is to address this)

TL;DR: Actions from `sink`s will be applied synchronously, causing an immediate render. Actions from `Workers` and `Subscriptions` will continue to be queued so they are not received synchronously.

The `sink` will be updated only be valid after `render` has completed, until the next `render` pass. This is true now, but the update will make it "fail fast" so it's obvious instead of silently dropping actions.

# Problem statement:

Events can be lost, due to the render pass happening (async) after the processing of an event. Results in a sink no longer being valid for a period before the UI has been provided a new sink.


# Current behavior:
```
(workflow render -> screen) -> (UI "onTap" bound to -> screen.onTap)
--- main thread yielded ---
(UI tap -> screen.onTap) -> workflow.Action.apply(.onTap)
--- main thread yielded ---
(workflow render -> screen) -> (UI "onTap" bound to -> screen.onTap)
```

## Detail of failure mode:

```
(workflow render -> screen) -> (UI "onTap" bound to -> screen.onTap)
--- main thread yielded ---
(UI tap -> screen.onTap) -> workflow.Action.apply(.onTap)
--- main thread yielded ---

--- NEXT RENDER PASS QUEUED ---
(UI tap -> screen.onTap) -> "into the ether" (the previous sink is no longer valid, specifically the signal behind it is not longer subscribed)
--- main thread yielded ---
(workflow render -> screen) -> (UI "onTap" bound to -> screen.onTap)
```

## Initial Design proposal:
```
(workflow render -> screen) -> (UI "onTap" bound to -> screen.onTap)
--- main thread yielded ---
(UI tap -> screen.onTap) -> workflow.Action.apply(.onTap) -> (workflow render -> screen) -> (UI "onTap" bound to -> screen.onTap)
```


# Details of actual subscription/wiring process:

## Visualization

```
      /--------\
      | Screen |
      \--------/
          ^
          |
          | Render (sync)
          |
/-------------------\
|                   | async /---------------\ sync /------\
|                   |<------| Subscriptions |<-----| Sink |
|                   |       \---------------/      \------/
|     Workflow      |
|                   |   async   /--------\
|                   |<----------| Worker |
|                   |           \--------/
\-------------------/
         ^
         |
         | Output (sync)
         |
/-------------------\
|                   |
|                   |
|       Child       | async  /-----\
|     Workflow      |<-------| ... |
|                   |        \-----/
|                   |
|                   |
\-------------------/
```

Render:

call `render` on the workflow with "RenderContext":
- create sinks (each sink creates a signal)
- create children (each child creates a signal for output)
- create workers (each worker has a signal producer created, which specifically has a signal. Current subscribed on the workflow scheduler)

(aside, the `SubtreeManager`'s `onUpdate` closure is wired to the `WorkflowNode` `handle` method (that processes actions)

`render` completes:
- context is invalidated. Calls to it will assert.
- childWorkflow `.onUpdate` closure is wired up to the subtree managers `.handle` function (which calls `.onUpdate`)
- childWorker `.onUpdate` closure is wired up to the subtree managers `.handle` function
- The event sources (ie: the signal's for all of the sinks, as well as any `subscribe` calls) are merged into a single signal, and mapped into an Output of "update"(`AnyWorkflowAction`)
    - The other type of `Output` is childDidUpdate, which will induce a render pass when it gets to the top of the tree.
- The merged signal of event sources is sent to the `childEvent` signal (which is a signal of signals). It's `flatMap`'d to `.latest`, so only one event will be handled from all the signals.
    - When any of the signal's emit, the `handle` function will be called (which calls the `.onUpdate` to get back to the WorkflowNode)

The `rendering` is returned from the subtree manager's "render".


# Use Cases

## General Design Goals/Constraints

- Only one event should be handled per render pass. Handling a single event must result in a new `render`.
    - This guarantees that the Action being handled is handled from an "expected" state.
- Thus, there are two limitations:
    - Multiple events cannot happen at the "same" time.
    - Events cannot be queued. (This is potentially debatable)

The two limitations are in some sense the exact same limitation (since if there is a queue and only the first is pulled off, there is only one event. If two calls happen synchronously two would still be queued, but again only the first is pulled off).

## UI originated Events

The callback invoked from UI to generate an event must be (conceptually) valid the next time the UI is able to generate an event (from UIKit).
- This means that if a new callback is handed to the UI on each render pass (and is only valid for a single event), the new callback must be provided before UIKit can generate a new UI event.
    - (as it stands now) If we fatal errored on a second event being send to a sink, we would currently have a race condition that would cause crashes. Instead, we are silently dropping events.
- Can we dispatch the event down to the workflow asynchronously when it's received?
    - No? If the UI thread yields after sending a single event, and then our queue gets priority, it would have to get back to the UI thread to update the screen. The UI thread could send another event.
- The sink should conceptually become "valid" as soon as `render` completes.
- An event update must immediately change the state and cause render to be called.

## Workers

A worker emits some action (ideally in the future) from the SignalProducer that is returned from `run`. The "ideally in the future" part is the complexity we need to handle:

1) The signal producer does something async, and then emits an action
    - This is trivial, and handled.
2) The signal producer is in the shape of `SignalProducer(value: <whatever>)` that emits immediately.
    - We need to handle this async, so either it must be subscribed on a scheduler that is not immediate, or not started until after `render` is completed.

## Subscriptions

1) A subscription emits a single event some time in the future.
    - This is trivial, and handled.
2) A subscription emits immediatley when it's subscribed.
    - Due to how Signal conceptually are, this shouldn't happen. However, it *could* with sink's (and it's been the case in the past where it does happen immediately, and it's dropped).
    - Back to "a sink should not be valid until `render` is completed.
3) A subscription emits two values synchronously.
    - This is not supported, see UI events above (again, worth revisiting)

## Child Workflows (Output events)

Child workflows are likely the most "easy" ones to handle (for output events). Because a child workflow can *only* generate an output during an `apply` call on an Action, they are just the originator of an event and it should be trivial to ensure they synchronously flow up the tree.

- Child has an action happen from some source.
- Emits and output
- Parent immediately received the output as an action and processes it, optionally emitting another output (and the loop continues)

This should be made clear in the subtree manager to guarantee it's the behavior. It definitely must not move schedulers, or we would have a partial tree update (and be missing calls to render).

# Queueing Events

As noted above, currently only a single event (action) is processed for each render loop. This can result in dropped events if (for instance) `send` is called twice on a sink synchronously. The first event is processed, and the second is dropped. This behavior can present with subscriptions and workers. Child workflows do not have this behavior, as only a single output can be emitted with each Action `apply`.

Adding queueing of events conceptually seems like a way to solve this, but it introduces (likely) "surprising" behavior when applying the actions.

## Example 1 (subscription):
1) `render` is called, and a signal is subscribed to, due to being in a "waiting" state (just a case in the state enum).
    -  The `apply` of the subscribed signal action assumes that the current state is `waiting`, due to it only being subscribed when that is the current state.
2) The signal emits two events, which are mapped into two actions.
    - The second action is queued.
3) The first action is applied. The `apply` updates the state to "not waiting". `render` occures, and the signal is *not* subscribed to.
4) The queued event from the previous render pass is applied. The state is no longer `waiting`, so the assumption of the state when applying the action is invalid.
    - The state is updated a second time, and potentially a behavior bug has happened.

The "solution" when writing this is to validate the expected state anytime an action is being applied. However, this increases the likelihood of bugs, as it may not be expected that a workflow would have to handle actions from a previous render loop.

# Proposal

*Discount multiple events during a single render pass. Queueing adds too much complexity when writing business logic.*

To ensure that events are not lost from the UI, as well as subscriptions and workers, the subtree manager will be updated to have a mix of synchronous and asynchronous behaviors depending on the event source.

- UI Events (from a `sink`). The sink type will no longer be backed by a subscription/signal, but instead will be a separate underlying type.
    - A `sink` will only become "valid" after the call to `render` has completed. This prevents an action being emitted during a `render` call, avoid reentrancy.
    - UI Events from a sink will be applied synchronously. They will immediately induce an update, and `render` will be called prior to returning from the `sink.send` call.
- Child output events will (continue) be applied synchronously.
- Subscriptions to signals we be done on a separate scheduler from the UI scheduler. This ensures that a `Signal` that emits immediately on `subscribe` will not be dropped.
- Workers that must be `run` will be subscribed on the same scheduler as subscriptions. This ensures that immediate events will not be dropped.

The handling of updates from events will be separated from how the events are injected into the  system. This provides the path for a `sink` event to be immediate, but workers and subscriptions to be asynchronous.

## Visualized proposal

```
      /--------\
      | Screen |
      \--------/
          ^
          |
          | Render (sync)
          |
/-------------------\
|                   |      /--------\    sync    /------\
|                   |      |        |<-----------| Sink |
|                   |      |        |            \------/
|                   |      |        |
|                   | sync |        | async /---------------\
|     Workflow      |<-----| Update |<------| Subscriptions |
|                   |      |        |       \---------------/
|                   |      |        |
|                   |      |        |   async   /--------\
|                   |      |        |<----------| Worker |
|                   |      \--------/           \--------/
\-------------------/           ^
                                |
                                |
                                | Output (sync)
                                |
                       /-------------------\
                       |                   |
                       |                   |
                       |       Child       | (as above) /-----\
                       |     Workflow      |<-----------| ... |
                       |                   |            \-----/
                       |                   |
                       |                   |
                       \-------------------/
```

**Comment by zach-klippenstein** [comment]
Thanks for doing such a thorough write-up! Couple questions:

1. How can an event handler emit immediately during a render pass, if it's backed by a sink and events are only sent on that sink by the consumer of the rendering?

2. How many schedulers do workflows typically have in play internally? Is it more than zero?

**Comment by davidapgar** [comment]
For (1), A sink exists during the call to `render` on a workflow (they are created with the `RenderContext`, ie: `RenderContext.makeSink(of: ActionType)`) - it would be considered a programmer error to call `sink.send()` while inside of `render` (but should be an obvious error, not silently dropped).

For (2), there is one scheduler in play aside from whatever the events are being dispatched on (generally the main thread). The scheduler is what subscriptions to signal (observables) and workers are observed on. The specific one guarantees that it will always queue, and never emit immediately.

**Comment by zach-klippenstein** [comment]
Do things blow up if someone writes a worker that explicitly emits on a separate scheduler?

**Comment by davidapgar** [comment]
For a worker, if someone emits on a separate scheduler, we listen to it on the one we want to handle the actions on - so we move it to the correct scheduler

**Comment by davidapgar** [comment]
Added wrinkle to this when testing these changes against their usage internally:

The PR https://github.com/square/workflow/pull/362 makes all actions from sinks come induce a render synchronously. In order to avoid a reentrant loop of actions causing repeated renders (ie: action -> render -> receive action -> render again, call stack keeps increasing), asserts were put in, and the sinks have to be "enabled" which happens after the rendering has been handed off.

The case where this breaks down:
A UITextField has a few ways it can send down updates:
1) target/action for `UIControlEventEditingChanged`
  - This case is ok - UIKit only sends a single event on each runloop, so it handles the queueing
2) `UIControlEventEditingDidEnd` and/or `UIControlEventEditingDidEndOnExit`, which is called when the text field resigns first responder.
  - This case is not ok. See below

An update to the Rendering (screen) may cause a view to be removed from the view hierarchy (eg: if it was the visible view in a back stack/nav controller, and a new one is being pushed). The causes `resignFirstResponder` to be called synchronously, which calls through from the action of `EditingDidEnd`. The method receiving this call calls `sink.send(.editingEnded)` or `sink.send(.textUpdated)`. This is a synchronous action, and would either be a) reentrant, or b) asserts with the defensive added about enabling sinks after render completes.

Initial thought: Allow queuing only for the period between `render` on the workflow completing, and `enableEvents` being called.
Can be done one of two ways:
1) Have the action be dispatched asynchronously (like with subscriptions or workers).
  - Has the risk/flaw of still having an event be dropped, as there is the case that a UIKit event (such as a pending keypress) could be queued that would not be sent to a valid sink, or that it would arrive "first", and the queued event would then be dropped.
2) Have the action be dispatched synchronously (immediately applied when `enableEvents` is called) inducing another render pass.
  - This is at least no longer reentrant, as the render pass has fully completed and the UI side has updated (so a workflow is not being updated in a partially applied state).
  - It *does* make handling of an action in WorkflowHost be effectively recursive in this case. There is the general concern of call stack depth/endless updates. It (at first glance) would be fairly "safe", as `enableEvents` can be the last thing called in the `handle(output:)` method - makes this akin to tail recursion (still recursive, but not reentrant - all other event/action handling is canceled at the start of the next render pass).

I *think* with (2), this covers the original case this issue was filed - the receiver of the screen synchronously gets to update the sinks it uses from a given event (so the lost keyboard entry example is "ok" - UIKit each time it gets back the runloop will give another keypress). It does add a "worse" behavior if the UI side always calls `sink.send` every time it gets a new screen - specifically, before the main runloop was yielded each time (and then the action immediately applied), as opposed to blocking the main runloop until the screen updates stop immediately sending actions. The prior behavior would have been just as "broken" (and endless update loop), but would present worse with it constantly blocking the main runloop.

Going to explore this some more, but welcome any additional thoughts.
